lib/plugins/aws/invokeLocal/index.js

```diff
@@ -150,41 +152,55 @@ class AwsInvokeLocal {
-      Number(this.serverless.service.provider.memorySize) ||
-      1024;
-
-    const lambdaDefaultEnvVars = {
-      LANG: 'en_US.UTF-8',
-      LD_LIBRARY_PATH:
-        '/usr/local/lib64/node-v4.3.x/lib:/lib64:/usr/lib64:/var/runtime:/var/runtime/lib:/var/task:/var/task/lib', // eslint-disable-line max-len
-      LAMBDA_TASK_ROOT: '/var/task',
-      LAMBDA_RUNTIME_DIR: '/var/runtime',
-      AWS_REGION: this.provider.getRegion(),
-      AWS_DEFAULT_REGION: this.provider.getRegion(),
-      AWS_LAMBDA_LOG_GROUP_NAME: this.provider.naming.getLogGroupName(lambdaName),
-      AWS_LAMBDA_LOG_STREAM_NAME: '2016/12/02/[$LATEST]f77ff5e4026c45bda9a9ebcec6bc9cad',
-      AWS_LAMBDA_FUNCTION_NAME: lambdaName,
-      AWS_LAMBDA_FUNCTION_MEMORY_SIZE: memorySize,
-      AWS_LAMBDA_FUNCTION_VERSION: '$LATEST',
-      NODE_PATH: '/var/runtime:/var/task:/var/runtime/node_modules',
-    };
+    return BbPromise.try(() => {
+      const lambdaName = this.options.functionObj.name;
+      const memorySize =
+        Number(this.options.functionObj.memorySize) ||
+        Number(this.serverless.service.provider.memorySize) ||
+        1024;
+
+      const lambdaDefaultEnvVars = {
+        LANG: 'en_US.UTF-8',
+        LD_LIBRARY_PATH:
+          '/usr/local/lib64/node-v4.3.x/lib:/lib64:/usr/lib64:/var/runtime:/var/runtime/lib:/var/task:/var/task/lib', // eslint-disable-line max-len
+        LAMBDA_TASK_ROOT: '/var/task',
+        LAMBDA_RUNTIME_DIR: '/var/runtime',
+        AWS_REGION: this.provider.getRegion(),
+        AWS_DEFAULT_REGION: this.provider.getRegion(),
+        AWS_LAMBDA_LOG_GROUP_NAME: this.provider.naming.getLogGroupName(lambdaName),
+        AWS_LAMBDA_LOG_STREAM_NAME: '2016/12/02/[$LATEST]f77ff5e4026c45bda9a9ebcec6bc9cad',
+        AWS_LAMBDA_FUNCTION_NAME: lambdaName,
+        AWS_LAMBDA_FUNCTION_MEMORY_SIZE: memorySize,
+        AWS_LAMBDA_FUNCTION_VERSION: '$LATEST',
+        NODE_PATH: '/var/runtime:/var/task:/var/runtime/node_modules',
+      };
 
-    const credentialEnvVars = this.getCredentialEnvVars();
+      const credentialEnvVars = this.getCredentialEnvVars();
 
-    // profile override from config
-    const profileOverride = this.provider.getProfile();
-    if (profileOverride) {
-      lambdaDefaultEnvVars.AWS_PROFILE = profileOverride;
-    }
+      // profile override from config
+      const profileOverride = this.provider.getProfile();
+      if (profileOverride) {
+        lambdaDefaultEnvVars.AWS_PROFILE = profileOverride;
+      }
 
-    const configuredEnvVars = this.getConfiguredEnvVars();
+      const configuredEnvVars = this.getConfiguredEnvVars();
 
-    _.merge(process.env, lambdaDefaultEnvVars, credentialEnvVars, configuredEnvVars);
+      const promises = _.entries(configuredEnvVars).map(([name, value]) => {
+        return this.resolveCfFunctionValue(this.provider, value)
+          .then(resolvedValue => {
+            configuredEnvVars[name] = resolvedValue;
+          })
+          .catch(e => {
```

Comment:
This should be a second callback to previous `then` (otherwise we're catching here eventual programmer error in _first_ then callback, and that can hide important issues)

Reply:
Ok, I updated accordingly.

Just for my knowledge, my understanding of promise chainable then/catch was indeed to avoid nesting then and catch and have all instructions on the same level.

How is 

```
return bool ? resolveAsync() : BbPromise.resolve(value)
  .then(value => {
    doSomething
  })
  .catch(e => {
    doSomething
  })
```

different from 

```
return bool ? resolveAsync().catch(e => {
    doSomething
  }) : BbPromise.resolve(value)
  .then(value => {
    doSomething
  })
```

Especially since I throw again in my catch statement. This serve as error context enriching, even if error is thrown in the `then` statement, it will benefit the user having, no ?

Reply:
@fredericbarthelet I was referring strictly to lines as commented (your comment seems to dive into some other area).

Note the important difference between the outcome of two below snippets:

```javascript
const configuredEnvVars = {}
const potentiallyErrorneousResolution = Promise.resole("successValue");
potentiallyErrorneousResolution.then(resolvedValue => {
  configuredEvnVars[name] = resolvedValue;
}).catch(e => {
  throw new Error("Could not resolve environment  variable")
});
```

```javascript
const configuredEnvVars = {}
const potentiallyErrorneousResolution = Promise.resole("successValue");
potentiallyErrorneousResolution .then(resolvedValue => {
  configuredEvnVars[name] = resolvedValue;
}, e => {
  throw new Error("Could not resolve environment  variable")
});
```

Unfortunately erroneous `then(onSuccess).catch(onFailure)` pattern is repeated by most of promise tutorials and you find in many docs. While _if/else_ logic is resembled correctly only by `then(onSuccess, onFailure)`

`catch(onFailure)` is nothing more than convenient shortcut for `then(null, onFailure)`, and ideally should be seen just as that.